### DISCUSSION TITLE
Does runtime persistence options thread safed?

### DISCUSSION THREAD

**Discussion by midnight-wonderer** [mongodb/mongoid#5770] [Q&A] [answered]
I just read about this neat feature about [Runtime Persistence Options](https://www.mongodb.com/docs/mongoid/master/reference/persistence-configuration/#runtime-persistence-options).

To recap, Mongoid allows

~~~Ruby
MyModel.with({ client: :query }) do |collection|
  collection.last
  # something else
end
~~~

to explicitly specify client usage for some operations.

During the operation, if some other thread accesses `MyModel`, does it use `:default` client or `:query` client?
It isn't obvious from [the code](https://github.com/mongodb/mongoid/blob/5237212c36d778e7bff3e53de1ad49058da2866d/lib/mongoid/clients/options.rb#L197); as the library seems to mutate the state to me.
I wonder if this is thread-safe.

Also, is there any other way I can utilize non-default clients?

**Comment by jamis** [comment, marked as answer]
Hello @midnight-wonderer! The persistence options are thread safe; they ultimately call `PersistenceContext#get_context` ([here](https://github.com/mongodb/mongoid/blob/5237212c36d778e7bff3e53de1ad49058da2866d/lib/mongoid/persistence_context.rb#L293-L296)), which fetches the information from thread-local storage.

By "non-default" clients, do you mean clients created ad-hoc, instead of declared in the mongoid config? The `client` option requires a client name, but you can declare a new client and assign it a name programmatically, like so:

~~~ruby
Mongoid::Clients.set(:client_name, client_instance)
~~~

In general, we encourage developers to use previously-declared clients for this, rather than ad-hoc clients. Use of ad-hoc clients can make it difficult to understand what the behavior of a particular query or update really is.
